validate.go

```diff
@@ -0,0 +1,63 @@
+package cid
+
+import (
+	"fmt"
+
+	mh "github.com/multiformats/go-multihash"
+)
+
+var ErrPossiblyInsecureHashFunction = fmt.Errorf("potentially insecure hash functions not allowed")
+var ErrBelowMinimumHashLength = fmt.Errorf("hashes must be at least bytes long")
+
+const minimumHashLength = 20
+
+var goodset = map[uint64]bool{
+	mh.SHA2_256:     true,
+	mh.SHA2_512:     true,
+	mh.SHA3_224:     true,
+	mh.SHA3_256:     true,
+	mh.SHA3_384:     true,
+	mh.SHA3_512:     true,
+	mh.SHAKE_256:    true,
+	mh.DBL_SHA2_256: true,
+	mh.KECCAK_224:   true,
+	mh.KECCAK_256:   true,
+	mh.KECCAK_384:   true,
+	mh.KECCAK_512:   true,
+	mh.ID:           true,
+
+	// insecure
+	mh.SHA1:    false,
+	mh.MURMUR3: false,
+}
+
+func IsGoodHash(code uint64) bool {
+	good, found := goodset[code]
+	if good {
+		return true
+	}
+
+	if !found {
+		if code >= mh.BLAKE2B_MIN+19 && code <= mh.BLAKE2B_MAX {
+			return true
+		}
+		if code >= mh.BLAKE2S_MIN+19 && code <= mh.BLAKE2S_MAX {
+			return true
+		}
+	}
+
+	return false
+}
+
+func ValidateCid(c *Cid) error {
+	pref := c.Prefix()
+	if !IsGoodHash(pref.MhType) {
+		return ErrPossiblyInsecureHashFunction
+	}
+
+	if pref.MhType != mh.ID && pref.MhLength < minimumHashLength {
```

Comment:
I'd rather check if the hash length >= the default. Alternatively, make `goodset` map hash function to minimum length (where we'd use -1 for ID).

Reply:
Hrm... I still want to allow some truncation. Ethereum truncates a lot of things down to 20 bytes (from 32). I can go with a mapping, but maybe default 20 and have a special cases for anything else?

Reply:
I think we should set this per hash as @Stebalien suggested to allow a little more flexibility.

Reply:
@whyrusleeping WTF? That's 80 bits of security! No!

Reply:
@Stebalien hold up, 8 * 20 is 160.

Reply:
@Stebalien the whole point of multiformats in general is to allow users to use what fits their needs best. We should continue using 256bit hashes for our important things, but shouldnt begrudge users who want to use 160bit hashes. I'm primarily concerned about preventing pre-image attacks, general collision birthday problem attacks are still a concern, but less so for now.

Reply:
So, I've come to agree on this. It still annoys me, but we don't have much of a choice.

---

If we ever need to set different minimum lengths for different hashes, we can change that later.

Reply:
Also, users can implement custom DAGServices for validating additional properties (we may want won for unixfs that has stricter requirements).